File: backend/app/schemas/ml_governance.py
import json
from datetime import datetime
from typing import Optional, List, Any, Dict
from pydantic import BaseModel, Field, validator
from app.models.ml_feedback import FeedbackType, ReviewStatus


class FeedbackCreate(BaseModel):
    event_id: Optional[int] = Field(None, description="Linked disaster event ID if applicable")
    hazard_type: str = Field(..., description="flood or landslide")
    feedback_type: FeedbackType = Field(..., description="Type of feedback")
    observed_outcome: Optional[str] = Field(None, description="What actually happened")
    notes: Optional[str] = Field(None, description="Additional notes from user")
    feature_snapshot: Optional[Dict[str, Any]] = Field(None, description="Snapshot of features used during prediction")
    provenance: Optional[str] = Field(None, description="Origin of the feedback (e.g. Citizen Report, Field Validation)")

    @validator("feature_snapshot", pre=True)
    def parse_snapshot(cls, v):
        if isinstance(v, str):
            try:
                return json.loads(v)
            except Exception:
                return {}
        return v


class FeedbackReviewRequest(BaseModel):
    review_status: ReviewStatus = Field(..., description="New status (e.g., ACCEPTED, REJECTED, UNDER_REVIEW)")
    notes: Optional[str] = Field(None, description="Reviewer notes")


class GroundTruthApprovalRequest(BaseModel):
    evidence_type: str = Field(..., description="Type of evidence (e.g., 'verified_field_report')")
    evidence_reference: str = Field(..., description="Reference link or ID for the evidence")
    notes: Optional[str] = Field(None, description="Approval notes")


class FeedbackResponse(BaseModel):
    id: int
    event_id: Optional[int]
    hazard_type: str
    feedback_type: FeedbackType
    review_status: ReviewStatus
    observed_outcome: Optional[str]
    notes: Optional[str]
    feature_snapshot: Optional[Dict[str, Any]]
    provenance: Optional[str]
    evidence_type: Optional[str]
    evidence_reference: Optional[str]
    exclusion_reason: Optional[str]
    submitted_by_id: Optional[int]
    submitted_at: datetime
    reviewed_by_id: Optional[int]
    reviewed_at: Optional[datetime]

    class Config:
        from_attributes = True

    @validator("feature_snapshot", pre=True)
    def parse_snapshot_out(cls, v):
        if isinstance(v, str):
            try:
                return json.loads(v)
            except Exception:
                return {}
        return v


class DatasetCandidateSummary(BaseModel):
    id: int
    dataset_version: str
    hazard_type: str
    status: str
    record_count: int
    positive_samples: int
    negative_samples: int
    feature_schema: List[str]
    geographic_coverage: Optional[str]
    temporal_coverage: Optional[str]
    methodology: Optional[str]
    exclusions: Optional[str]
    
    included_count: int = 0
    excluded_count: int = 0
    duplicate_count: int = 0
    leakage_count: int = 0
    unknown_provenance_count: int = 0

    created_at: datetime
    created_by_id: Optional[int]


class CandidateExportRecord(BaseModel):
    feedback_id: int
    event_id: Optional[int]
    hazard_type: str
    feedback_type: str
    observed_outcome: Optional[str]
    feature_snapshot: Dict[str, Any]
    provenance: Optional[str]
    evidence_type: Optional[str]
    evidence_reference: Optional[str]

File: frontend/src/lib/api/types/ml-governance.ts
export type FeedbackType = "confirmed_event" | "false_positive" | "false_negative" | "uncertain" | "data_correction";
export type ReviewStatus = "submitted" | "under_review" | "accepted" | "rejected" | "approved_ground_truth";

export interface FeedbackCreate {
  event_id?: number | null;
  hazard_type: string;
  feedback_type: FeedbackType;
  observed_outcome?: string | null;
  notes?: string | null;
  feature_snapshot?: Record<string, any> | null;
  provenance?: string | null;
}

export interface FeedbackResponse {
  id: number;
  event_id?: number | null;
  hazard_type: string;
  feedback_type: FeedbackType;
  review_status: ReviewStatus;
  observed_outcome?: string | null;
  notes?: string | null;
  feature_snapshot?: Record<string, any> | null;
  provenance?: string | null;
  evidence_type?: string | null;
  evidence_reference?: string | null;
  exclusion_reason?: string | null;
  submitted_by_id?: number | null;
  submitted_at: string;
  reviewed_by_id?: number | null;
  reviewed_at?: string | null;
}

export interface FeedbackReviewRequest {
  review_status: ReviewStatus;
  notes?: string | null;
}

export interface DatasetCandidateSummary {
  id: number;
  dataset_version: string;
  hazard_type: string;
  status: string;
  record_count: number;
  positive_samples: number;
  negative_samples: number;
  feature_schema: string[];
  geographic_coverage: string;
  temporal_coverage: string;
  methodology: string;
  exclusions?: string;
  included_count: number;
  excluded_count: number;
  duplicate_count: number;
  leakage_count: number;
  unknown_provenance_count: number;
  created_at: string;
  created_by_id?: number;
}

export interface CandidateExportRecord {
  feedback_id: number;
  event_id?: number | null;
  hazard_type: string;
  feedback_type: string;
  observed_outcome?: string | null;
  feature_snapshot: Record<string, any>;
  provenance?: string | null;
}
